90zfsbootmenu/zfsbootmenu-lib.sh

```diff
@@ -1136,14 +1137,112 @@ be_is_locked() {
   return 1
 }
 
+# arg1: BE key source
+# prints: value of org.zfsbootmenu:keysource for the BE, if defined
+# returns: 0 iff the value is defined and not empty
+
+be_keysource() {
+  local fs keysource
+  fs="${1}"
+
+  if ! keysrc="$( zfs get -H -o value org.zfsbootmenu:keysource "${fs}" )"; then
+    zdebug "failed to read org.zfsbootmenu:keysource on ${fs}"
+    echo ""
+    return 1
+  fi
+
+  if [ "x${keysrc}" = "x-" ] || [ -z "${keysrc}" ]; then
+    echo ""
+    return 1;
+  fi
+
+  echo "${keysrc}"
+  return 0
+}
+
+# arg1: ZFS filesystem
+# arg2: key location
+# prints: nothing
+# returns: 0 iff a key was cached
+
+cache_key() {
+  local fs keylocation keyfile keydir keysrc keycache mutex mnt ret
+  fs="${1}"
+  keylocation="${2}"
+
+  # Strip scheme if it exists
+  keyfile="${keylocation#file://}"
+  # Make file relative to root
+  keyfile="${keyfile#/}"
+
+  if [ "x${keyfile}" = "x${keylocation}" ] || [ -z "${keyfile}" ]; then
+    # No change or no file => keylocation is not a file => nothing to cache
+    return 1;
+  fi
+
+  # Make sure a key source is defined
+  if ! keysrc="$( be_keysource "${fs}" )"; then
+    return 1
+  fi
+
+  keycache="${BASE}/.keys/${keysrc}"
+  mutex="${keycache}/.cachemutex"
+
+  if [ -e "${mutex}" ]; then
+    # Attempting to load key source could lead to infinite recursion
+    # Break the chain by only attempting to cache if not previously attempted
+    zdebug "will not repeat cache attempt for ${keysrc}:${keyfile}"
+    return 1
+  fi
+
+  # Populate the cache if possible
+  zdebug "attempting to cache ${keysrc}:${keyfile}"
+
+  if ! mkdir -p "${keycache}"; then
+    return 1
+  fi
+
+  # Make sure to touch the cache mutex
+  : > "${mutex}"
+  if [ ! -f "${mutex}" ]; then
+    zdebug "failed to acquire mutex ${mutex}"
+    return 1
+  fi
+
+  if ! load_key "${keysrc}" || ! mnt="$( mount_zfs "${keysrc}" )"; then
```

Comment:
Can we break this pair of actions up a little bit so that each one can be individually tested and the failure subsequently logged as a `zerror`?